formatter/formatter_test.go

```diff
@@ -0,0 +1,222 @@
+				Input: mustReadFile(path.Join(testSourceDir, f.Name())),
+			})
+			if gqlErr != nil {
+				t.Fatal(gqlErr)
+			}
+
+			// exec format
+			var buf bytes.Buffer
+			err := formatter.NewFormatter(&buf).FormatSchemaDocument(doc)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// validity check
+			_, gqlErr = parser.ParseSchema(&ast.Source{
+				Name:  f.Name(),
+				Input: buf.String(),
+			})
+			if gqlErr != nil {
+				t.Log(buf.String())
+				t.Fatal(gqlErr)
+			}
+
+			return buf.Bytes()
+		},
+	})
+}
+
+func TestFormatter_FormatQueryDocument(t *testing.T) {
+	const testSourceDir = "./testdata/source/query"
+	const testBaselineDir = "./testdata/baseline/FormatQueryDocument"
+
+	executeGoldenTesting(t, &goldenConfig{
+		SourceDir: testSourceDir,
+		BaselineFileName: func(cfg *goldenConfig, f os.FileInfo) string {
+			return path.Join(testBaselineDir, f.Name())
+		},
+		Run: func(t *testing.T, cfg *goldenConfig, f os.FileInfo) []byte {
+			// load stuff
+			doc, gqlErr := parser.ParseQuery(&ast.Source{
+				Name:  f.Name(),
+				Input: mustReadFile(path.Join(testSourceDir, f.Name())),
+			})
+			if gqlErr != nil {
+				t.Fatal(gqlErr)
+			}
+
+			// exec format
+			var buf bytes.Buffer
+			err := formatter.NewFormatter(&buf).FormatQueryDocument(doc)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			// validity check
+			_, gqlErr = parser.ParseQuery(&ast.Source{
+				Name:  f.Name(),
+				Input: buf.String(),
+			})
+			if gqlErr != nil {
+				t.Log(buf.String())
+				t.Fatal(gqlErr)
+			}
+
+			return buf.Bytes()
+		},
+	})
+}
+
+type goldenConfig struct {
+	SourceDir        string
+	IsTarget         func(f os.FileInfo) bool
+	BaselineFileName func(cfg *goldenConfig, f os.FileInfo) string
+	Run              func(t *testing.T, cfg *goldenConfig, f os.FileInfo) []byte
+}
+
+func executeGoldenTesting(t *testing.T, cfg *goldenConfig) {
+	t.Helper()
+
+	if cfg.IsTarget == nil {
+		cfg.IsTarget = func(f os.FileInfo) bool {
+			return !f.IsDir()
+		}
+	}
+	if cfg.BaselineFileName == nil {
+		t.Fatal("BaselineFileName function is required")
+	}
+	if cfg.Run == nil {
+		t.Fatal("Run function is required")
+	}
+
+	fs, err := ioutil.ReadDir(cfg.SourceDir)
+	if err != nil {
+		t.Fatal(fs)
+	}
+
+	for _, f := range fs {
+		if f.IsDir() {
+			continue
+		}
+		f := f
+
+		t.Run(f.Name(), func(t *testing.T) {
+			result := cfg.Run(t, cfg, f)
+
+			expectedFilePath := cfg.BaselineFileName(cfg, f)
+			expected, err := ioutil.ReadFile(expectedFilePath)
+			if os.IsNotExist(err) {
+				err = os.MkdirAll(path.Dir(expectedFilePath), 0755)
+				if err != nil {
+					t.Fatal(err)
+				}
+				err = ioutil.WriteFile(expectedFilePath, result, 0444)
+				if err != nil {
+					t.Fatal(err)
+				}
+				return
+			} else if err != nil {
+				t.Fatal(err)
+			}
+
+			if bytes.Equal(expected, result) {
+				return
+			}
+
+			t.Logf("if you want to accept new result. rm %s", expectedFilePath)
+
+			if utf8.Valid(expected) {
+				diff := difflib.UnifiedDiff{
```

Comment:
what about `go test ./... -u` to update?

save adding an extra dependency?

Reply:
It's means create new `-u` flag to test case some thing like https://speakerdeck.com/mitchellh/advanced-testing-with-go?slide=20 ?
I didn't know it. It is seems good. I'll try it later.

so, I add new dependency `github.com/pmezard/go-difflib/difflib` to go.mod. Is it bad?

Reply:
Looks like testify brings in difflib anyway.

Reply:
make sense. The code uses testify only now, and exec `go mod tidy`. this PR doesn't contains any change in go.mod & go.sum 😉